### PR TITLE
[reopen][feat] GitHub App integration #23

### LINKED ISSUE #15: [feat] Instructor can install GithubApp
# Acceptance criteria
- [x] There's a documented (in `/docs` of the repo) process by which the developer of ProjOrganic can set up a Github App
- [x] There is a defined process by which an instructor can authorize the GitHub App to invite users to the GitHub Organization.
- [ ] There is a demonstration of how a student on the course roster can log in and see a list of the courses they are students for and can click join to join any one of them.

### PR DESCRIPTION
**Pull request by yuxiaolejs**
In this PR, I added code to enable organic to function as a Github App. By doing so, instructors now can link our github app to their course, which in current stage, can invite their students to the course organization automatically

## Pre-req
There's no hard pre-req for this branch, but merging #68 first will make the testing process of this branch much easier.

## Feature

Backend now is integrated with github app, you can interact with it using:
- `GitHubUserApi` bean is used to authenticate as the user is logged in.
- `GitHubApp` bean is used to authenticate as a github app.
- `GitHubToken` bean exposes the Oauth token of the current user, you can use it to call github apis manually.

From now on, this system works with Github App instead of Oauth app, and I have provided a documentation on setting in up in the [docs/oauth.md](docs/oauth.md) **You MUST have a private key configured to make it work!**

One new page is added, and it looks like this:

image

One static page is added, and it looks like this:

image

You can find it [here](https://ucsb-cs156-s24.github.io/proj-organic-s24-5pm-3/prs/23/storybook/?path=/story/pages-courses-coursejoinpage--default) in the storybook, but it's not quite working since it depends on back end data.

## Deployment

I have a staged build for you to test it. You can find it here: https://organic.dokku.tianleyu.com/

Dokku dev deployment available here: https://organic-yuxiaolejs-dev.dokku-11.cs.ucsb.edu

Story book is now available here, but please note that it does not work since it tries to communicate with the backend (which is not there in storybook) https://ucsb-cs156-s24.github.io/proj-organic-s24-5pm-3/prs/46/storybook/?path=/story/pages-courses-coursejoinpage--default

## Limitations
Now this checks roster, but students can't be populated for testing since not a single endpoint for doing that is ready. Working on that. Please see PR #68, which addresses this problem. It is recommended to merge #68 before this one if time permits.

## Additional changes
In addition to the changes mentioned in #23, this PR re-factored the prompt for instructors to link the github org with their courses. Also, I deprecated the package `com.tianleyu.github` and copied everything inside of it over to this project under `edu.ucsb.cs156.github`

Closes #15

### CODE REVIEW

**Review by 30912hyl — COMMENTED** [review]
that's a lot of changed files...would it be better/possible to split this into smaller PRs?

**Comment by yuxiaolejs** [comment]
> that's a lot of changed files...would it be better/possible to split this into smaller PRs?

Not really... They are dependent on each other

**Comment by 30912hyl** [comment]
LGTM

**Review by 30912hyl — DISMISSED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -4,8 +4,8 @@
 
 | Type | Link       | 
 |------|------------|
-| prod | <https://organic.dokku-00.cs.ucsb.edu/>     | 
-| qa   | <https://organic-qa.dokku-00.cs.ucsb.edu/>  | 
+| prod | <https://organic.dokku-11.cs.ucsb.edu/>     | 
+| qa   | <https://organic-qa.dokku-11.cs.ucsb.edu/>  |
```
Comment (pconrad) [inline README.md]: Please revert this.  This is intended as documentation of the production links.

File: frontend/src/tests/pages/CoursesJoinPage.test.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            expect(screen.getByText(/UCSB/)).toBeInTheDocument();
+            expect(screen.getByText(/F23/)).toBeInTheDocument();
+            expect(screen.getByTestId("CourseJoin-school")).toBeInTheDocument();
+            expect(screen.getByTestId("CourseJoin-school")).toHaveTextContent("Course school: UCSB");
+            // restoreConsole();
+        });
+
+        test("Do the job when click join", async () => {
+            axiosMock.onPost("/api/courses/join").reply(202, {});
+            // const restoreConsole = mockConsole();
```
Comment (pconrad) [inline frontend/src/tests/pages/CoursesJoinPage.test.js]: commented out code.

File: frontend/src/tests/pages/CoursesJoinPage.test.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            await screen.findByText("Join")
+            axiosMock.history.get.forEach(req => console.log(req.url));
+            expect(axiosMock.history.get.length).toBeGreaterThanOrEqual(1);
+            expect(screen.getByText("Join")).toBeInTheDocument();
+            expect(screen.getByTestId("CourseJoin-school")).toBeInTheDocument();
+            expect(screen.getByText(/UCSB/)).toBeInTheDocument();
+            expect(screen.getByText(/F23/)).toBeInTheDocument();
+            expect(screen.getByTestId("CourseJoin-school")).toBeInTheDocument();
+            expect(screen.getByTestId("CourseJoin-school")).toHaveTextContent("Course school: UCSB");
+            // restoreConsole();
```
Comment (pconrad) [inline frontend/src/tests/pages/CoursesJoinPage.test.js]: commented out code

File: frontend/src/tests/pages/CoursesJoinPage.test.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            axiosMock.reset();
+            axiosMock.resetHistory();
+            axiosMock.onGet("/api/currentUser").reply(200, apiCurrentUserFixtures.userOnly);
+            axiosMock.onGet("/api/systemInfo").reply(200, systemInfoFixtures.showingNeither);
+            axiosMock.onGet("/api/courses/join", { params: { id: 17 } }).reply(200, coursesFixtures.oneCourse);
+        });
+
+        const queryClient = new QueryClient();
+        test("Load data with backend info", async () => {
+            // const restoreConsole = mockConsole();
```
Comment (pconrad) [inline frontend/src/tests/pages/CoursesJoinPage.test.js]: commented out code

File: frontend/src/tests/pages/CoursesJoinPage.test.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        const axiosMock = new AxiosMockAdapter(axios);
+        const queryClient = new QueryClient();
+        beforeEach(() => {
+            axiosMock.reset();
+            axiosMock.resetHistory();
+            axiosMock.onGet("/api/currentUser").reply(200, apiCurrentUserFixtures.userOnly);
+            axiosMock.onGet("/api/systemInfo").reply(200, systemInfoFixtures.showingNeither);
+        });
+        test("Load data with backend info", async () => {
+            // const restoreConsole = mockConsole();
```
Comment (pconrad) [inline frontend/src/tests/pages/CoursesJoinPage.test.js]: commented out code

File: frontend/src/tests/pages/CoursesJoinPage.test.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            render(
+                <QueryClientProvider client={queryClient}>
+                    <MemoryRouter>
+                        <CourseJoinPage />
+                    </MemoryRouter>
+                </QueryClientProvider>
+            );
+            await screen.findByText("Loading")
+            // restoreConsole();
```
Comment (pconrad) [inline frontend/src/tests/pages/CoursesJoinPage.test.js]: commented out code

**Review by pconrad — CHANGES_REQUESTED** [review]
There are some commented out lines of code.

Please fix these; then the code looks good.  I will experiment with it on my own dokku deployment, and see if we can get it to a mergeable state.

This is really solid good work, and will earn extra credit points for the team. I'm hopeful, though that the team can get to 100 points without it :-)

**Pull request closed without merging**